!pip install transformers  
установим spaCy (run in terminal/prompt)  
!{sys.executable} -m pip install spacy
Загрузим spaCy's  'en' Model  
!{sys.executable} -m spacy download en

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подготовка" data-toc-modified-id="Подготовка-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подготовка</a></span><ul class="toc-item"><li><span><a href="#Загрузка-данных" data-toc-modified-id="Загрузка-данных-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Загрузка данных</a></span></li><li><span><a href="#Подготовка-функций" data-toc-modified-id="Подготовка-функций-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Подготовка функций</a></span></li><li><span><a href="#Подготовка-данных-для-моделей" data-toc-modified-id="Подготовка-данных-для-моделей-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Подготовка данных для моделей</a></span><ul class="toc-item"><li><span><a href="#Модели-с-векторизацией-Bag-of-words-и-TF-IDF" data-toc-modified-id="Модели-с-векторизацией-Bag-of-words-и-TF-IDF-1.3.1"><span class="toc-item-num">1.3.1&nbsp;&nbsp;</span>Модели с векторизацией Bag of words и TF-IDF</a></span></li><li><span><a href="#Модели-с-векторизацией-BERT" data-toc-modified-id="Модели-с-векторизацией-BERT-1.3.2"><span class="toc-item-num">1.3.2&nbsp;&nbsp;</span>Модели с векторизацией BERT</a></span></li></ul></li></ul></li><li><span><a href="#Обучение-и-выбор-моделей" data-toc-modified-id="Обучение-и-выбор-моделей-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Обучение и выбор моделей</a></span><ul class="toc-item"><li><span><a href="#Векторизация-Bag-of-words" data-toc-modified-id="Векторизация-Bag-of-words-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Векторизация Bag of words</a></span><ul class="toc-item"><li><span><a href="#Логистическая-регрессия-(bag-of-words)" data-toc-modified-id="Логистическая-регрессия-(bag-of-words)-2.1.1"><span class="toc-item-num">2.1.1&nbsp;&nbsp;</span>Логистическая регрессия (bag of words)</a></span></li><li><span><a href="#Модель-на-основе-случайного-леса-(bag-of-words)" data-toc-modified-id="Модель-на-основе-случайного-леса-(bag-of-words)-2.1.2"><span class="toc-item-num">2.1.2&nbsp;&nbsp;</span>Модель на основе случайного леса (bag of words)</a></span></li></ul></li><li><span><a href="#Векторизация-TF-IDF-(Term-Frequency–Inverse-Document-Frequency)" data-toc-modified-id="Векторизация-TF-IDF-(Term-Frequency–Inverse-Document-Frequency)-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Векторизация TF-IDF (Term Frequency–Inverse Document Frequency)</a></span><ul class="toc-item"><li><span><a href="#Логистическая-регрессия-(TF-IDF)" data-toc-modified-id="Логистическая-регрессия-(TF-IDF)-2.2.1"><span class="toc-item-num">2.2.1&nbsp;&nbsp;</span>Логистическая регрессия (TF-IDF)</a></span></li><li><span><a href="#Модель-на-основе-случайного-леса-(TF-IDF)" data-toc-modified-id="Модель-на-основе-случайного-леса-(TF-IDF)-2.2.2"><span class="toc-item-num">2.2.2&nbsp;&nbsp;</span>Модель на основе случайного леса (TF-IDF)</a></span></li></ul></li><li><span><a href="#Векторизация-BERT" data-toc-modified-id="Векторизация-BERT-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Векторизация BERT</a></span><ul class="toc-item"><li><span><a href="#Логистическая-регрессия-(BERT)" data-toc-modified-id="Логистическая-регрессия-(BERT)-2.3.1"><span class="toc-item-num">2.3.1&nbsp;&nbsp;</span>Логистическая регрессия (BERT)</a></span></li><li><span><a href="#Модель-на-основе-случайного-леса-(BERT)" data-toc-modified-id="Модель-на-основе-случайного-леса-(BERT)-2.3.2"><span class="toc-item-num">2.3.2&nbsp;&nbsp;</span>Модель на основе случайного леса (BERT)</a></span></li></ul></li><li><span><a href="#Выбор-и-тестирование-модели" data-toc-modified-id="Выбор-и-тестирование-модели-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Выбор и тестирование модели</a></span></li></ul></li><li><span><a href="#Выводы" data-toc-modified-id="Выводы-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Выводы</a></span></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li></ul></div>

# Классификация комментариев с BERT

Интернет-магазин «Викишоп» запускает новый сервис. Теперь пользователи могут редактировать и дополнять описания товаров, как в вики-сообществах. То есть клиенты предлагают свои правки и комментируют изменения других. Магазину нужен инструмент, который будет искать токсичные комментарии и отправлять их на модерацию. 

Обучите модель классифицировать комментарии на позитивные и негативные. В вашем распоряжении набор данных с разметкой о токсичности правок.

Постройте модель со значением метрики качества *F1* не меньше 0.75. 

**Инструкция по выполнению проекта**

1. Загрузите и подготовьте данные.
2. Обучите разные модели. 
3. Сделайте выводы.

Для выполнения проекта применять *BERT* необязательно, но вы можете попробовать.

**Описание данных**

Данные находятся в файле `toxic_comments.csv`. Столбец *text* в нём содержит текст комментария, а *toxic* — целевой признак.

 <div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> 
    
Вижу твое добавленное описание проекта. Молодец! Это поможет тебе расставлять акценты в выводах

In [74]:
# Импортируем необходимые библиотеки
import numpy as np
import pandas as pd
from tqdm import notebook
from tqdm import tqdm
import re
import gc
import joblib

import nltk
from nltk.corpus import stopwords as nltk_stopwords
nltk.download('stopwords')

# Для лемматизации слов
import sys
import spacy

# Для создания собственных трансформеров
from sklearn.base import BaseEstimator, TransformerMixin

from sklearn.pipeline import Pipeline

from sklearn.feature_extraction.text import (CountVectorizer,
                                             TfidfVectorizer)
# Масштабирование признаков в разреженных матрицах
from sklearn.preprocessing import MaxAbsScaler

from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier

from sklearn.model_selection import (train_test_split,
                                     GridSearchCV) 
from sklearn.metrics import f1_score

# Для работы с BERT
import torch
import transformers

# Константы
RANDOM_STATE = 5

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\Дмитрий\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


## Подготовка

### Загрузка данных

In [2]:
# Загрузим данные
try:
    df = pd.read_csv('toxic_comments.csv')
except:
    df = pd.read_csv('/datasets/toxic_comments.csv')

In [3]:
# Рассмотрим общие сведения
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 159292 entries, 0 to 159291
Data columns (total 3 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   Unnamed: 0  159292 non-null  int64 
 1   text        159292 non-null  object
 2   toxic       159292 non-null  int64 
dtypes: int64(2), object(1)
memory usage: 3.6+ MB


In [4]:
# Рассмотрим первые строки
df.head()

Unnamed: 0                                               text  toxic
0           0  Explanation\nWhy the edits made under my usern...      0
1           1  D'aww! He matches this background colour I'm s...      0
2           2  Hey man, I'm really not trying to edit war. It...      0
3           3  "\nMore\nI can't make any real suggestions on ...      0
4           4  You, sir, are my hero. Any chance you remember...      0

Данные содержат:
- столбец `text`с комментариями на английском языке,
- столбец `toxic`, содержащий значение 1 для токсичного комментария. **Целевой признак**,
- столбец `Unnamed: 0`, не несущий какой-либо информации.

In [5]:
# Удалим стобец `Unnamed: 0`.
df = df.drop(['Unnamed: 0'], axis=1)

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 159292 entries, 0 to 159291
Data columns (total 2 columns):
 #   Column  Non-Null Count   Dtype 
---  ------  --------------   ----- 
 0   text    159292 non-null  object
 1   toxic   159292 non-null  int64 
dtypes: int64(1), object(1)
memory usage: 2.4+ MB


In [7]:
# Посмотрим балансировку классов
df['toxic'].value_counts(normalize=True)

toxic
0    0.898388
1    0.101612
Name: proportion, dtype: float64

Имеется разбалансировка классов, поэтому в моделях будем применять взвешивание классов.

### Подготовка функций

In [8]:
# Комментарии на английском языке, поэтому загрузим стоп-слова для английского языка
stopwords = set(nltk_stopwords.words('english'))
stop_words = list(stopwords)

In [9]:
# Для выполнения лемматизации
# инициализируем модель spacy 'для английского, оставив только компонент, необходимый для лемматизации
nlp = spacy.load('en_core_web_sm', disable=['parser', 'ner'])

In [10]:
# Определим функцию для предобработки текста
def text_preprocessor(text):
    # Приведём все символы к нижнему регистру
    out_text = text.lower()

    # Удалим стоп-слова
    out_text = out_text.split()
    clean_tokens = tuple(
                   map(lambda x: '' if x in stop_words else x, out_text)
    )
    out_text = ' '.join(clean_tokens)

    # Удалим все символы, которые не являются буквами латинского алфавита
    out_text = re.sub(r'[^a-zA-Z]', ' ', out_text)

    # Выполним лемматизацию
    # Parse the sentence using the loaded 'en' model object `nlp`
    doc = nlp(out_text)
    # Extract the lemma for each token and join
    out_text = ' '.join([token.lemma_ for token in doc])

    # Все одиночные символы, имеющие пробелы с обеих сторон, заменим пробелом
    out_text = re.sub(r'\s+[a-z]\s+', ' ', out_text)
    
    # Все последовательности пробелов заменим пробелом
    out_text = ' '.join(out_text.split())
    
    return out_text

In [11]:
# Определим трансформер для векторизации Bag of words
class bow_transformer(BaseEstimator, TransformerMixin):
    def __init__(self, max_df):
        self.max_df = max_df
        self._estimator = CountVectorizer(encoding='utf8',
                                          preprocessor=None,
                                          lowercase=False,
                                          stop_words=None,
                                          max_df=self.max_df)
    def fit(self, X, y=None):
        _ = self._estimator.fit(X['lemm_text'])
        return self

    def transform(self, X):
        X_out = self._estimator.transform(X['lemm_text'])
        return X_out

In [12]:
# Определим трансформер для векторизации TF-IDF
class tfidf_transformer(BaseEstimator, TransformerMixin):
    def __init__(self, max_df, norm):
        self.max_df = max_df
        self.norm = norm
        self._estimator = TfidfVectorizer(encoding='utf8',
                                          preprocessor=None,
                                          lowercase=False,
                                          stop_words=None,
                                          max_df=self.max_df,
                                          norm=self.norm)
    def fit(self, X, y=None):
        _ = self._estimator.fit(X['lemm_text'])
        return self

    def transform(self, X):
        X_out = self._estimator.transform(X['lemm_text'])
        return X_out

In [13]:
# Масштабирование элементов разряженной матрицы
scaler = MaxAbsScaler()

In [69]:
# Модель логистической регрессии
lr_clf = LogisticRegression(random_state=RANDOM_STATE, solver='saga', class_weight='balanced')

In [15]:
# Модель на основе случайного леса
rf_clf = RandomForestClassifier(n_jobs=-1, random_state=RANDOM_STATE, class_weight='balanced')

### Подготовка данных для моделей

#### Модели с векторизацией Bag of words и TF-IDF

In [16]:
# Подготовка текстов для векторизации Bag of words и TF-IDF
df['lemm_text'] = df['text']

# Для контроля прогресса будем использовать progress_apply
tqdm.pandas()
df['lemm_text'] = df['text'].progress_apply(text_preprocessor)

# Сохраним результат предобработки текста в файл
df.to_csv('comments_with_lemmatized_text.csv', index=False)

100%|█████████████████████████████████████████████████████████████████████████| 159292/159292 [08:33<00:00, 310.20it/s]


In [17]:
# Рассмотрим данные
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 159292 entries, 0 to 159291
Data columns (total 3 columns):
 #   Column     Non-Null Count   Dtype 
---  ------     --------------   ----- 
 0   text       159292 non-null  object
 1   toxic      159292 non-null  int64 
 2   lemm_text  159292 non-null  object
dtypes: int64(1), object(2)
memory usage: 3.6+ MB


#### Модели с векторизацией BERT

Прежде чем перевести тексты в векторы, подготовим их. У BERT есть собственный токенизатор. Это инструмент, который разбивает и преобразует исходные тексты в список токенов, которые есть в словаре. Лемматизация не требуется.  
Для работы возьмём предобученную модель с ресурса https://huggingface.co/unitary/toxic-bert

Будем использовать:
- модель (файл pytorch_model.bin),
- файл конфигурации модели (config.json),
- словарь для токенизации (файл vocab.txt).

In [19]:
# Инициализируем токенизатор как объект класса BertTokenizer().
# Передадим ему аргумент vocab_file — это файл со словарём, на котором обучалась модель.
# Он может быть, например, в текстовом формате (txt)
tokenizer = transformers.BertTokenizer(do_lower_case=True,
    vocab_file='toxic_bert/vocab.txt')

# Преобразуем текст в номера токенов из словаря методом encode()
# Для корректной работы модели мы указали аргумент add_special_tokens (англ. «добавить специальные токены»),
# равный True. Это значит, что к любому преобразуемому тексту добавляется токен начала (101) и токен конца текста (102). 
# Для уменьшения количества токенов применим truncation
tokenized = df['text'].progress_apply(
    lambda x: tokenizer.encode(x, add_special_tokens=True, max_length=512, truncation='longest_first'))

# Применим метод padding (англ. «отступ»), чтобы после токенизации длины исходных текстов в корпусе были равными.
# Только при таком условии будет работать модель BERT. Пусть стандартной длиной вектора n будет длина наибольшего
# во всём датасете вектора. Остальные векторы дополним нулями
max_len = 0
for i in tokenized.values:
    if len(i) > max_len:
        max_len = len(i)

padded = np.array([i + [0]*(max_len - len(i)) for i in tokenized.values])

#Теперь поясним модели, что нули не несут значимой информации. Это нужно для компоненты модели,
# которая называется «внимание» (англ. attention). Отбросим эти токены и «создадим маску» для действительно
# важных токенов, то есть укажем нулевые и не нулевые значения:
attention_mask = np.where(padded != 0, 1, 0)

100%|█████████████████████████████████████████████████████████████████████████| 159292/159292 [03:09<00:00, 841.54it/s]


In [20]:
padded.shape

(159292, 512)

In [21]:
attention_mask.shape

(159292, 512)

In [22]:
# освободим оперативную память
del tokenized
del df
gc.collect()

1690

In [23]:
# Инициализируем саму модель класса BertModel.
# Передадим ей файл с предобученной моделью и конфигурацией:
config = transformers.BertConfig.from_json_file(
    'toxic_bert/config.json')
model = transformers.BertModel.from_pretrained(
    'toxic_bert/pytorch_model.bin', config=config)

Преобразуем тексты в эмбеддинги. Для наглядного контроля прогресса подключим библиотеку tqdm (араб. taqadum, تقدّم, «прогресс»). В Jupyter применим функцию notebook() из этой библиотеки.

In [35]:
# Эмбеддинги моделью BERT будем создавать батчами. Чтобы хватило оперативной памяти, сделаем размер батча небольшим
batch_size = 100
embeddings = []
k = 0
# Сделаем цикл по батчам. Отображать прогресс будет функция notebook()
for i in notebook.tqdm(range(round(np.ceil(padded.shape[0]/batch_size)))):
        # Преобразуем данные в формат тензоров (англ. tensor) — многомерных векторов в библиотеке torch.
        # Тип данных LongTensor (англ. «длинный тензор») хранит числа в «длинном формате», то есть выделяет
        # на каждое число 64 бита.
        # Преобразуем данные
        batch = torch.LongTensor(padded[batch_size*i:batch_size*(i+1)]) 
        # Преобразуем маску
        attention_mask_batch = torch.LongTensor(attention_mask[batch_size*i:batch_size*(i+1)])
        # Чтобы получить эмбеддинги для батча, передадим модели данные и маску
        # Для ускорения вычисления функцией no_grad() (англ. no gradient, «нет градиента») в
        # библиотеке torch укажем, что градиенты не нужны: модель BERT обучать не будем.
        with torch.no_grad():
            batch_embeddings = model(batch, attention_mask=attention_mask_batch)
        # Из полученного тензора извлечём нужные элементы (для каждого текста выберем эмбеддинг первого токена,
        # соответствующего началу текста) и добавим в список всех эмбеддингов
        # преобразуем элементы методом numpy() к типу numpy.array
        embeddings.append(batch_embeddings[0][:,0,:].numpy())
        # Каждые 5000 текстов будем освобождать оперативную память, сохраняя результат в файл
        if len(embeddings) == 50:
            print(f'saving file embeddings_{k}.csv. indexes = {k*50*batch_size}:{(k+1)*50*batch_size-1}')
            pd.DataFrame(np.concatenate(embeddings)).to_csv(f'embeddings_{k}.csv',
                                                            index=False)
            k+=1
            del embeddings
            gc.collect()
            embeddings = []
print(f'saving file embeddings_{k}.csv. indexes = {k*50*batch_size}:{padded.shape[0]-1}')
pd.DataFrame(np.concatenate(embeddings)).to_csv(f'embeddings_{k}.csv',
                                                            index=False)

  0%|          | 0/1593 [00:00<?, ?it/s]

saving file embeddings_0.csv. indexes = 0:4999
saving file embeddings_1.csv. indexes = 5000:9999
saving file embeddings_2.csv. indexes = 10000:14999
saving file embeddings_3.csv. indexes = 15000:19999
saving file embeddings_4.csv. indexes = 20000:24999
saving file embeddings_5.csv. indexes = 25000:29999
saving file embeddings_6.csv. indexes = 30000:34999
saving file embeddings_7.csv. indexes = 35000:39999
saving file embeddings_8.csv. indexes = 40000:44999
saving file embeddings_9.csv. indexes = 45000:49999
saving file embeddings_10.csv. indexes = 50000:54999
saving file embeddings_11.csv. indexes = 55000:59999
saving file embeddings_12.csv. indexes = 60000:64999
saving file embeddings_13.csv. indexes = 65000:69999
saving file embeddings_14.csv. indexes = 70000:74999
saving file embeddings_15.csv. indexes = 75000:79999
saving file embeddings_16.csv. indexes = 80000:84999
saving file embeddings_17.csv. indexes = 85000:89999
saving file embeddings_18.csv. indexes = 90000:94999
saving fil

In [36]:
# Очистим оперативную память
del embeddings
del config
del model
gc.collect()

18

In [37]:
# Соберём эмбеддинги в один файл
embeddings = []
for i in range(k+1):
    embeddings.append(pd.read_csv(f'embeddings_{i}.csv').values)
pd.DataFrame(np.concatenate(embeddings)).to_csv(f'embeddings_all.csv',
                                                            index=False)

In [38]:
# очистим оперативную память
del embeddings
gc.collect()

0

## Обучение и выбор моделей

### Векторизация Bag of words

In [50]:
# Загрузим данные
df = pd.read_csv('comments_with_lemmatized_text.csv')

In [59]:
# Рассмотрим данные
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 159292 entries, 0 to 159291
Data columns (total 3 columns):
 #   Column     Non-Null Count   Dtype 
---  ------     --------------   ----- 
 0   text       159292 non-null  object
 1   toxic      159292 non-null  int64 
 2   lemm_text  159269 non-null  object
dtypes: int64(1), object(2)
memory usage: 3.6+ MB


In [61]:
# Рассмотрим пропуски
df[df['lemm_text'].isna()]

text  toxic lemm_text
3983                    From here\n\nFrom here 160.80.2.8      0       NaN
4475    1993\n\n1994\n\n1995\n\n1996\n\n1997\n\n1998\n...      0       NaN
6289                                 193.61.111.53  15:00      0       NaN
8824                             What is I 78.146.102.144      0       NaN
10193                                      64.86.141.133"      0       NaN
17280                                ~ \n\n68.193.147.157      0       NaN
23485                               between 1991 and 1999      0       NaN
38743                                       88.104.31.21"      0       NaN
46969                               and then 70.8.194.249      0       NaN
52336                                              14:53,      0       NaN
53679                       92.24.199.233|92.24.199.233]]      0       NaN
61644                            "\n\n 199.209.144.211  "      0       NaN
79643                        Some 1 should 216.237.210.43      0       NaN
82541                              "\n '''''' 2010/2013 "      0       NaN
100596                              [24 and 25, after 23]      0       NaN
119018                                                """      1       NaN
119697                                 What 24.119.163.71      0       NaN
124235                    What what what what is this ???      0       NaN
137400                                             == """      0       NaN
148595                               she did 76.122.79.82      0       NaN
151107                      10 - 2010 04 08 to 2010 05 12      0       NaN
152543                       SAME FOR THIS 166.137.240.20      0       NaN
153041  which is OVER 9000 OVER 9000 OVER 9000 OVER 90...      0       NaN

In [62]:
# пропуски заполним значением необработанного текста, добавив слово text
df.loc[df['lemm_text'].isna(), 'lemm_text'] = 'text' + df.loc[df['lemm_text'].isna(), 'text']

In [63]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 159292 entries, 0 to 159291
Data columns (total 3 columns):
 #   Column     Non-Null Count   Dtype 
---  ------     --------------   ----- 
 0   text       159292 non-null  object
 1   toxic      159292 non-null  int64 
 2   lemm_text  159292 non-null  object
dtypes: int64(1), object(2)
memory usage: 3.6+ MB


In [64]:
# Разделим признаки
X = pd.DataFrame(df['lemm_text'].copy(), columns=['lemm_text'])
y = df['toxic'].copy()

In [65]:
# Разделим выборки
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=RANDOM_STATE, stratify=y
)

In [66]:
print(f'Обучающая выборка {X_train.shape[0]} ({X_train.shape[0]/X.shape[0]:.2f}), \
тестовая выборка {X_test.shape[0]} ({X_test.shape[0]/X.shape[0]:.2f})')

Обучающая выборка 127433 (0.80), тестовая выборка 31859 (0.20)


In [67]:
# Очистим оперативную память
del df
del X
gc.collect()

7728

#### Логистическая регрессия (bag of words)

In [90]:
lr_bow_pl = Pipeline(steps=[('vect', bow_transformer(max_df=1.0)),
                            ('scaling', scaler),
                            ('clf', lr_clf)], verbose=True)

param_grid_lr_bow = [{'vect__max_df': [2, 10, 100],
                      'clf__max_iter': [2000],
                      'clf__C': range(1, 21),
                      'clf__penalty':['l2']
                      }
                     ]
gs_lr_bow_pl = GridSearchCV(
                           lr_bow_pl,
                           param_grid=param_grid_lr_bow,
                           scoring='f1',
                           cv=4,
                           n_jobs=-1,
                           verbose=3,
                           error_score='raise'
)

gs_lr_bow_pl.fit(X_train, y_train)

gs_lr_bow_best_score = gs_lr_bow_pl.best_score_
gs_lr_bow_best_params = gs_lr_bow_pl.best_params_

# лучшее значение f1 на кросс-валидации
print(f'best_score: {gs_lr_bow_best_score}')
# лучшие гиперпараметры
print(f'best_params: {gs_lr_bow_best_params}')
# Сохраним модель
_ = joblib.dump(gs_lr_bow_pl.best_estimator_, 'gs_lr_bow_pl.pkl', compress=0)

Fitting 4 folds for each of 60 candidates, totalling 240 fits
[Pipeline] .............. (step 1 of 3) Processing vect, total=   4.5s
[Pipeline] ........... (step 2 of 3) Processing scaling, total=   0.1s
[Pipeline] ............... (step 3 of 3) Processing clf, total= 1.1min
best_score: 0.543321815153645
best_params: {'clf__C': 6, 'clf__max_iter': 2000, 'clf__penalty': 'l2', 'vect__max_df': 2}


#### Модель на основе случайного леса (bag of words)

In [91]:
rf_bow_pl = Pipeline(steps=[('vect', bow_transformer(max_df=1.0)),
                            ('clf', rf_clf)], verbose=True)

param_grid_rf_bow = [{'vect__max_df': [2, 10, 100],
                      'clf__n_estimators': [10, 50, 100],
                      'clf__max_depth': [5, 20, None]
                      }
                     ]
gs_rf_bow_pl = GridSearchCV(
                           rf_bow_pl,
                           param_grid=param_grid_rf_bow,
                           scoring='f1',
                           cv=4,
                           n_jobs=-1,
                           verbose=3,
                           error_score='raise'
)

gs_rf_bow_pl.fit(X_train, y_train)

gs_rf_bow_best_score = gs_rf_bow_pl.best_score_
gs_rf_bow_best_params = gs_rf_bow_pl.best_params_

# лучшее значение f1 на кросс-валидации
print(f'best_score: {gs_rf_bow_best_score}')
# лучшие гиперпараметры
print(f'best_params: {gs_rf_bow_best_params}')
# Сохраним модель
_ = joblib.dump(gs_rf_bow_pl.best_estimator_, 'gs_rf_bow_pl.pkl', compress=0)

Fitting 4 folds for each of 27 candidates, totalling 108 fits
[Pipeline] .............. (step 1 of 2) Processing vect, total=   4.6s
[Pipeline] ............... (step 2 of 2) Processing clf, total= 1.4min
best_score: 0.6050004185533085
best_params: {'clf__max_depth': None, 'clf__n_estimators': 100, 'vect__max_df': 2}


### Векторизация TF-IDF (Term Frequency–Inverse Document Frequency)

#### Логистическая регрессия (TF-IDF)

In [92]:
lr_tfidf_pl = Pipeline(steps=[('vect', tfidf_transformer(max_df=1.0, norm='l2')),
                              ('scaling', scaler),
                              ('clf', lr_clf)], verbose=True)

param_grid_lr_tfidf = [{'vect__max_df': [2, 10, 100],
                        'vect__norm':['l1', 'l2'],
                        'clf__max_iter': [2000],
                        'clf__C': range(1, 21),
                        'clf__penalty':['l2']
                       }
                      ]
gs_lr_tfidf_pl = GridSearchCV(
                            lr_tfidf_pl,
                            param_grid=param_grid_lr_tfidf,
                            scoring='f1',
                            cv=4,
                            n_jobs=-1,
                            verbose=3,
                            error_score='raise'
)

gs_lr_tfidf_pl.fit(X_train, y_train)

gs_lr_tfidf_best_score = gs_lr_tfidf_pl.best_score_
gs_lr_tfidf_best_params = gs_lr_tfidf_pl.best_params_

# лучшее значение f1 на кросс-валидации
print(f'best_score: {gs_lr_tfidf_best_score}')
# лучшие гиперпараметры
print(f'best_params: {gs_lr_tfidf_best_params}')
# Сохраним модель
_ = joblib.dump(gs_lr_tfidf_pl.best_estimator_, 'gs_lr_tfidf_pl.pkl', compress=0)

Fitting 4 folds for each of 120 candidates, totalling 480 fits
[Pipeline] .............. (step 1 of 3) Processing vect, total=   4.8s
[Pipeline] ........... (step 2 of 3) Processing scaling, total=   0.1s
[Pipeline] ............... (step 3 of 3) Processing clf, total=  15.4s
best_score: 0.7614821022533786
best_params: {'clf__C': 4, 'clf__max_iter': 2000, 'clf__penalty': 'l2', 'vect__max_df': 2, 'vect__norm': 'l1'}


#### Модель на основе случайного леса (TF-IDF)

In [93]:
rf_tfidf_pl = Pipeline(steps=[('vect', tfidf_transformer(max_df=1.0, norm='l2')),
                              ('clf', rf_clf)], verbose=True)

param_grid_rf_tfidf = [{'vect__max_df': [2, 10, 100],
                        'vect__norm':['l1', 'l2'],
                        'clf__n_estimators': [10, 50, 100],
                        'clf__max_depth': [5, 20, None]
                        }
                       ]
gs_rf_tfidf_pl = GridSearchCV(
                            rf_tfidf_pl,
                            param_grid=param_grid_rf_tfidf,
                            scoring='f1',
                            cv=4,
                            n_jobs=-1,
                            verbose=3,
                            error_score='raise'
)

gs_rf_tfidf_pl.fit(X_train, y_train)

gs_rf_tfidf_best_score = gs_rf_tfidf_pl.best_score_
gs_rf_tfidf_best_params = gs_rf_tfidf_pl.best_params_

# лучшее значение f1 на кросс-валидации
print(f'best_score: {gs_rf_tfidf_best_score}')
# лучшие гиперпараметры
print(f'best_params: {gs_rf_tfidf_best_params}')
# Сохраним модель
_ = joblib.dump(gs_rf_tfidf_pl.best_estimator_, 'gs_rf_tfidf_pl.pkl', compress=0)

Fitting 4 folds for each of 54 candidates, totalling 216 fits
[Pipeline] .............. (step 1 of 2) Processing vect, total=   4.8s
[Pipeline] ............... (step 2 of 2) Processing clf, total= 1.1min
best_score: 0.6053570563350497
best_params: {'clf__max_depth': None, 'clf__n_estimators': 100, 'vect__max_df': 2, 'vect__norm': 'l1'}


### Векторизация BERT

In [94]:
# Загрузим данные и разделим признаки, а потом освободим оперативную память
X_bert = pd.read_csv('embeddings_all.csv')

In [95]:
# Для экономии памяти преобразуем данные
X_bert = X_bert.astype('float32')

In [96]:
# Разделим выборки
X_train_bert, X_test_bert, y_train_bert, y_test_bert = train_test_split(
    X_bert, y, test_size=0.2, random_state=RANDOM_STATE)

In [97]:
print(f'Обучающая выборка {X_train_bert.shape[0]} ({X_train_bert.shape[0]/X_bert.shape[0]:.2f}), \
тестовая выборка {X_test_bert.shape[0]} ({X_test_bert.shape[0]/X_bert.shape[0]:.2f})')

Обучающая выборка 127433 (0.80), тестовая выборка 31859 (0.20)


In [98]:
# Очистим оперативную память
del X_bert
gc.collect()

201

#### Логистическая регрессия (BERT)

In [99]:
param_grid_lr_clf = {'max_iter': [2000],
                     'C': range(1, 21),
                     'penalty':['l2']
                     }
gs_lr_clf = GridSearchCV(
                            lr_clf,
                            param_grid=param_grid_lr_clf,
                            scoring='f1',
                            cv=4,
                            n_jobs=-1,
                            pre_dispatch=12, # для экономии оперативной памяти
                            verbose=3,
                            error_score='raise'
)

gs_lr_clf.fit(X_train_bert, y_train_bert)

gs_lr_clf_best_score = gs_lr_clf.best_score_
gs_lr_clf_best_params = gs_lr_clf.best_params_

# лучшее значение f1 на кросс-валидации
print(f'best_score: {gs_lr_clf_best_score}')
# лучшие гиперпараметры
print(f'best_params: {gs_lr_clf_best_params}')
# Сохраним модель
_ = joblib.dump(gs_lr_clf.best_estimator_, 'gs_lr_clf.pkl', compress=0)

Fitting 4 folds for each of 20 candidates, totalling 80 fits
best_score: 0.9246989833178894
best_params: {'C': 7, 'max_iter': 2000, 'penalty': 'l2'}


#### Модель на основе случайного леса (BERT)

In [100]:
param_grid_rf_clf = {'n_estimators': [10, 50, 100],
                        'max_depth': [5, 20, None]
                        }
gs_rf_clf = GridSearchCV(
                         rf_clf,
                         param_grid=param_grid_rf_clf,
                         scoring='f1',
                         cv=4,
                         n_jobs=-1,
                         pre_dispatch=12, # для экономии оперативной памяти
                         verbose=3,
                         error_score='raise'
)

gs_rf_clf.fit(X_train_bert, y_train_bert)

gs_rf_clf_best_score = gs_rf_clf.best_score_
gs_rf_clf_best_params = gs_rf_clf.best_params_

# лучшее значение f1 на кросс-валидации
print(f'best_score: {gs_rf_clf_best_score}')
# лучшие гиперпараметры
print(f'best_params: {gs_rf_clf_best_params}')
# Сохраним модель
_ = joblib.dump(gs_rf_clf.best_estimator_, 'gs_rf_clf.pkl', compress=0)

Fitting 4 folds for each of 9 candidates, totalling 36 fits
best_score: 0.9434044153885639
best_params: {'max_depth': 20, 'n_estimators': 100}


### Выбор и тестирование модели

Наилучшие результаты показала модель на основе случайного леса (гиперпараметры 'max_depth': 20, 'n_estimators': 100) с векторизацией текстов с использованием модели toxic bert. Проверим качество модели на тестовой выборке.

In [102]:
rf_bert_predict = gs_rf_clf.best_estimator_.predict(X_test_bert)
print(f'best_RandomForest_bert_test_f1: {f1_score(y_test_bert, rf_bert_predict):.3f}')

best_RandomForest_bert_test_f1: 0.949


## Выводы

В ходе работы выполнено:
- загрузка данных,
- подготовка исходных текстов для моделей с применением Bag of words, TF-IDF и для модели с применением BERT,
- определён способ предобработки текста (создан свой предобработчик),
- моделирование на основе логистической регрессии в трёх вариантах (с применением Bag of words, TF-IDF, BERT),
- моделирование на основе случайного леса в трёх вариантах (с применением Bag of words, TF-IDF, BERT),
- для обработки данных применён пайплайн, включающий трансформер собственной разработки, который позволяет изменять параметры векторизации текстов.


В ходе экспериментов удалось добиться максимального значения f1 0.949 на тестовой выборке, что соответствует заданным требованиям (f1 не менее 0.75). Наилучший результат показала модель на основе случайного леса с применением векторизации текстов моделью toxic bert. Гиперпараметры определены по итогам кросс-валидации.


**Рекомендуемые настройки для использования:**
- векторизация моделью toxic bert,
- модель на основе случайного леса (гиперпараметры max_depth: 20, n_estimators: 100, class_weight='balanced').